host/optee/linux/enclave.c

```diff
@@ -0,0 +1,472 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/defs.h>
+#include <openenclave/bits/safemath.h>
+#include <openenclave/edger8r/host.h>
+#include <openenclave/internal/calls.h>
+#include <openenclave/internal/defs.h>
+#include <openenclave/internal/raise.h>
+
+#include "../../calls.h"
+#include "../../ocalls.h"
+#include "enclave.h"
+
+#define PARAM_TYPES_IN_OUT      \
```

Comment:
Is this really all Linux-specific?

Reply:
No, but there's a "but".

The GP TEE Client API specification states that the API may be implemented either in non-secure user-mode or in non-secure kernel mode.

In Linux, the API is implemented by `libteec` in user-mode. On Windows, the API is implemented in kernel-mode in the OP-TEE miniport driver and a custom interface is exposed to user-mode (`CallOpteeCommand`).

So to answer your question, the logic in question here belongs in `CallOpteeCommand` on Windows. This, in turn, makes me realize that the latter must also change, in addition to the driver.

We should consider exposing the TEE Client API in user-mode on Windows, too.

Reply:
As a follow-up, if there is anything we can share with Windows, I'll refactor this file at that time.